src/agent/src/watcher.rs

```diff
@@ -0,0 +1,487 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use tokio::fs;
+use tokio::sync::Mutex;
+use tokio::task;
+use tokio::time::{self, Duration};
+
+use anyhow::{ensure, Context, Result};
+use async_recursion::async_recursion;
+use nix::mount::{umount, MsFlags};
+use slog::{error, Logger};
+
+use crate::mount::BareMount;
+use crate::protocols::agent::Storage;
+
+/// The maximum number of file system entries agent will watch for each mount.
+const MAX_ENTRIES_PER_STORAGE: usize = 8;
+
+/// How often to check for modified files.
+const WATCH_INTERVAL_SECS: u64 = 2;
+
+/// Destination path for tmpfs
+const WATCH_MOUNT_POINT_PATH: &str = "/run/kata-containers/shared/containers/watchable/";
+
+/// Represents a single file entry to be watched.
+#[derive(Debug, Clone)]
+struct File {
+    file: PathBuf,
+    modified: SystemTime,
+}
+
+/// Represents a single storage entry (may have multiple files to watch).
+#[derive(Default, Debug, Clone)]
+struct Entry {
+    source: PathBuf,
+    mount_point: PathBuf,
+    files: Vec<File>,
+}
+
+impl Entry {
+    async fn new(storage: Storage) -> Result<Entry> {
+        let source = PathBuf::from(&storage.source);
+        let mut mount_point = PathBuf::from(&storage.mount_point);
+
+        if source.is_file() && mount_point.is_dir() {
+            let filename = source.file_name().with_context(|| {
+                format!("Failed to extract file name from {}", source.display())
+            })?;
+            mount_point = mount_point.join(filename);
+        }
+
+        let mut entry = Entry::default();
+        entry.add_path(&source).await?;
+
+        entry.source = source;
+        entry.mount_point = mount_point;
+
+        Ok(entry)
+    }
+
+    /// Adds file entries from the given path
+    #[async_recursion]
+    async fn add_path(&mut self, path: &Path) -> Result<()> {
+        if path.is_file() {
+            self.add_file(path)
+                .await
+                .with_context(|| format!("Failed to add file {}", path.display()))?;
+        } else {
+            let mut entries = fs::read_dir(path)
+                .await
+                .with_context(|| format!("Failed to read dir: {}", path.display()))?;
+
+            while let Some(entry) = entries.next_entry().await? {
+                self.add_path(entry.path().as_path()).await?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Adds a single file path entry
+    async fn add_file(&mut self, file: &Path) -> Result<()> {
+        ensure!(
+            self.files.len() <= MAX_ENTRIES_PER_STORAGE,
+            "Too many file system entries to watch (must be < {})",
+            MAX_ENTRIES_PER_STORAGE
+        );
+
+        let entry = File {
+            file: file.to_path_buf(),
+            // Set minimum value to perform initial copy during next check
+            modified: SystemTime::UNIX_EPOCH,
```

Comment:
Won't this cause all files to be copied the first time `check()` is called? Why not just use the real file modification time here?

Reply:
We want the files to be copied over when we are setting up this "watch"